dotnet/src/MicroFocus.FAS.AdapterSdkSchema.SchemaObjectBuilders/StringDictionarySchemaObjectBuilderImpl.cs

```diff
@@ -0,0 +1,253 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace MicroFocus.FAS.AdapterSdkSchema.SchemaObjectBuilders
+{
+    internal sealed class StringDictionarySchemaObjectBuilderImpl : ISchemaObjectBuilder
+    {
+        private readonly Dictionary<string, IList<string>> _metadata;
+        private readonly IJsonStringBuilder _jsonStringBuilder;
+        private readonly string _prefix;
+
+        public StringDictionarySchemaObjectBuilderImpl(
+            Dictionary<string, IList<string>> metadata,
+            IJsonStringBuilder jsonStringBuilder
+        ) : this(metadata, jsonStringBuilder, "")
+        {
+        }
+
+        private StringDictionarySchemaObjectBuilderImpl(
+            Dictionary<string, IList<string>> metadata,
+            IJsonStringBuilder jsonStringBuilder,
+            string prefix
+        )
+        {
+            _metadata = metadata;
+            _jsonStringBuilder = jsonStringBuilder;
+            _prefix = prefix;
+        }
+
+        public void AddBooleanFieldValue(IField field, bool value)
+        {
+            AddStringFieldValueImpl(field, value.ToString());
+        }
+
+        public void AddDoubleFieldValue(IField field, double value)
+        {
+            AddStringFieldValueImpl(field, value.ToString());
+        }
+
+        public void AddInstantFieldValue(IField field, DateTime value)
+        {
+            AddStringFieldValueImpl(field, DateTimeFunctions.ToEpochSecondsString(value));
+        }
+
+        public void AddIntegerFieldValue(IField field, int value)
+        {
+            AddStringFieldValueImpl(field, value.ToString());
+        }
+
+        public void AddLongFieldValue(IField field, long value)
+        {
+            AddStringFieldValueImpl(field, value.ToString());
+        }
+
+        public void AddStringFieldValue(IField field, string value)
+        {
+            AddStringFieldValueImpl(field, value);
+        }
+
+        public void ClearField(IField field)
+        {
+            string metadataKey = GetMetadataKey(field);
+
+            if (field.ObjectEncoding == ObjectEncoding.Flattened)
+            {
+                List<string> keysToRemove = _metadata.Keys.Where(key => key.StartsWith(metadataKey + "_")).ToList();
+                keysToRemove.ForEach(key => _metadata.Remove(key));
+            }
+            else
+            {
+                _metadata.Remove(metadataKey);
+            }
+        }
+
+        public void SetBooleanFieldValue(IField field, bool value)
+        {
+            _metadata.Add(GetMetadataKey(field), new[] { value.ToString() });
```

Comment:
You're using `.Add()` here but that will fail if there is already a value present.